### PR TITLE
Add 2D grid with bilinear interpolation

### PR DESCRIPTION
**Pull request by sethrj**
This is needed for the simple Seltzer-Berger energy distribution tables.

### CODE REVIEW

File: src/physics/grid/TwodGridCalculator.i.hh [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        const real_type lower_val = grids_[ax][idx[ax]];
+        const real_type upper_val = grids_[ax][idx[ax] + 1];
+        frac[ax] = (inp[ax] - lower_val) / (upper_val - lower_val);
+    }
+
+    // Interpolate using the surorunding four values.
+    return (1 - frac[X]) * (1 - frac[Y]) * this->at(idx[X], idx[Y])
+           + (frac[X]) * (1 - frac[Y]) * this->at(idx[X] + 1, idx[Y])
+           + (1 - frac[X]) * (frac[Y]) * this->at(idx[X], idx[Y] + 1)
+           + (frac[X]) * (frac[Y]) * this->at(idx[X] + 1, idx[Y] + 1);
```
Comment (pcanal) [inline src/physics/grid/TwodGridCalculator.i.hh:53]: This layout is useful for readability consider
```suggestion
    // clang-format off
    return (1 - frac[X]) * (1 - frac[Y]) * this->at(idx[X], idx[Y])
           + (frac[X]) * (1 - frac[Y]) * this->at(idx[X] + 1, idx[Y])
           + (1 - frac[X]) * (frac[Y]) * this->at(idx[X], idx[Y] + 1)
           + (frac[X]) * (frac[Y]) * this->at(idx[X] + 1, idx[Y] + 1);
    // clang-format on
```

File: src/physics/grid/TwodGridInterface.hh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    //! True if assigned and valid
+    explicit CELER_FUNCTION operator bool() const
+    {
+        return x.size() >= 2 && y.size() >= 2
+               && values.size() == x.size() * y.size();
+    }
+
+    //! Get the data location for a specified x-y coordinate.
+    CELER_FUNCTION ItemId<real_type> at(size_type x, size_type y) const
```
Comment (pcanal) [inline src/physics/grid/TwodGridInterface.hh:33]: On some  compiler you might get a warning about the parameter shadowing the data members ...

File: test/physics/grid/TwodGridCalculator.test.cc [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        }
+        grid_data_.values = build.insert_back(values.begin(), values.end());
+        EXPECT_EQ(grid_data_.values.front(), grid_data_.at(0, 0));
+
+        CELER_ENSURE(grid_data_);
+        ref_ = values_;
+    }
+
+    // Bilinear function of (x, y) should exactly reproduce with interpolation
+    real_type f(real_type x, real_type y) const
```
Comment (pcanal) [inline test/physics/grid/TwodGridCalculator.test.cc:57]:
```suggestion
    real_type expected(real_type x, real_type y) const
```
to make the purpose of the function a tiny bit clearer in the using code.

File: src/physics/grid/TwodGridInterface.hh (hunk 2 above)
Comment (sethrj) [reply src/physics/grid/TwodGridInterface.hh:33]: Good point.

**Review by pcanal — APPROVED** [review]

**Pull request merged**